server/src/main/scala/org/http4s/server/middleware/RequestLogger.scala

```diff
@@ -0,0 +1,29 @@
+package org.http4s
+package server
+package middleware
+
+import fs2._
+import org.log4s._
+import cats.implicits._
+import fs2.interop.cats._
+
+/**
+  * Simple Middleware for Logging Requests As They Are Processed
+  */
+object RequestLogger {
+  private[this] val logger = getLogger
+
+  def apply(service: HttpService)(implicit strategy: Strategy): HttpService = Service.lift{ req =>
+    Stream(req)
+      .observe(requestLogSink)
+      .evalMap[Task, Task, MaybeResponse]{req => service(req)}
+      .runFoldMap(identity)
+  }
+
+  def requestLogSink: Sink[Task, Request] = stream => {
+    stream
+      .flatMap(_.body)
+      .through(fs2.text.utf8Decode)
```

Comment:
Pertaining to failure, that is why this happens in a side sink. This observes the request but I believe won't fail the request it is observing.(Not an expert and needs tests to confirm)

As for charset validation. Thats another area for expanded work. We could build charset pipes for rendering similar to this one although they might be better contributed to fs2. However as the next shift for fs2 involves `cats-effect` and some brand new structures would probably be better internally and then shifted to theirs after 1.0